[Feature] Persist plagiarism results

philippbauch opened the pull request:
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I added multiple integration tests (Jest) related to the features (with a high test coverage)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
Currently, there are three REST endpoints to start the automated plagiarism detection for modeling, programming, and text exercises:

* `GET /modeling-exercises/{exerciseId}/check-plagiarism`
* `GET /programming-exercises/{exerciseId}/check-plagiarism`
* `GET /text-exercises/{exerciseId}/check-plagiarism`

The `/check-plagiarism` endpoints for programming and text exercises return a `TextPlagiarismResult` and the `/check-plagiarism` endpoint for modeling exercises returns a `ModelingPlagiarismResult`.

However, the plagiarism detection results are not persisted on the Artemis server and get lost if the instructor refreshes the browser page. 

The goal of this PR is to persist plagiarism detection results to avoid unnecessary reruns. Instructors should have access to the latest plagiarism results only. They can rerun the automated plagiarism detection at any time to generate new results.

### Description
First, this PR introduces a new `PlagiarismService` that offers CRUD functionality for plagiarism results using the `PlagiarismResultRepository` and `PlagiarismComparisonRepository`.

Second, the `/check-plagiarism` route handlers now use the new `PlagiarismService` to store the latest result in the database before returning them to the client.
 
Third, we added three new REST endpoints to the server to access the latest plagiarism results for modeling, programming, and text exercises respectively:

* `GET /modeling-exercises/{exerciseId}/plagiarism-result`
* `GET /programming-exercises/{exerciseId}/plagiarism-result`
* `GET /text-exercises/{exerciseId}/plagiarism-result`

Last, we added a new REST endpoint to update the status of a single PlagiarismComparison. We also adapted the client to call this endpoint whenever the user clicks _Confirm_ or _Deny_ from the plagiarism inspector (see Screenshot).

### Steps for Testing
Make sure to test the following for modeling, programming, and text exercises. You'll new a few similar submissions for each exercise type to properly test this PR.

1. Log in to Artemis
2. Navigate to Course Administration
3. Go to the Exercises of a course
4. Go to the details page of a (modeling | programming | text) exercise with similar submissions
5. Click 'Check Plagiarism'

You're now on the Plagiarism Detection page. Make sure the following works:

- The plagiarism detection page should load the latest plagiarism detection result on page load (if any exists; otherwise empty page)
- Click 'Detect plagiarism' to start a new plagiarism detection run and receive results
- Highlighting of similarities for text and programming exercises should work
- Clicking _Confirm_ or _Deny_ correctly updates the status of a single comparison
- The updated status of a comparison is persisted and loaded correctly after refreshing the page
- Clicking 'Rerun plagiarism detection' creates a new result that overwrites the previous one. Especially check that the status of each comparison is reset

### Test Coverage
<img width="997" alt="Screenshot 2021-03-23 at 15 31 15" src="https://user-images.githubusercontent.com/21973908/112162988-deb6d980-8bec-11eb-9384-c491aeaea418.png">


### Screenshots
<img width="1165" alt="Screenshot 2021-03-16 at 09 18 52" src="https://user-images.githubusercontent.com/21973908/111277337-a5083080-8638-11eb-8ed2-8777919ae182.png">

krusche commented on src/main/java/de/tum/in/www1/artemis/domain/plagiarism/PlagiarismResult.java:
```diff
@@ -18,19 +50,25 @@
+    @JsonIdentityReference(alwaysAsId = true)
+    @JsonProperty("exerciseId")
+    @ManyToOne
+    protected Exercise exercise;
```
It would be perfectly find to load the exercise here and to use a simpler mechanism

krusche commented on src/main/java/de/tum/in/www1/artemis/domain/plagiarism/PlagiarismComparison.java:
```diff
@@ -1,30 +1,58 @@
-     * First submission involved in this comparison.
+     * First submission compared.
      */
+    @ManyToOne(targetEntity = PlagiarismSubmission.class, cascade = CascadeType.ALL)
```
CascadeType.ALL does not make sense in ManyToOne, because we might get situations where multiple PlagiarismComparisons reference the same  PlagiarismSubmission and then deleting only one 
PlagiarismComparison should potentially not delete the PlagiarismSubmission which is still references in another object

krusche replied:
maybe it is enough to document this case better, because we only allow to delete the container object and then all child objects should be deleted as well

krusche commented on src/main/java/de/tum/in/www1/artemis/domain/plagiarism/PlagiarismSubmission.java:
```diff
@@ -5,20 +5,32 @@
+    /**
+     * ID of the related submission.
+     */
+    private long submissionId;
```
improve documentation with the rationale you just told me on Skype ;-)

derLalla commented on src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java:
```diff
@@ -1218,28 +1264,34 @@ else if (!authCheckService.isAtLeastTeachingAssistantForExercise(programmingExer
 
+        Optional<PlagiarismResult> optionalPreviousResult = plagiarismService.getPlagiarismResult(programmingExercise);
+
         TextPlagiarismResult result = programmingExerciseExportService.checkPlagiarism(exerciseId, similarityThreshold, minimumScore);
 
-        log.info("Check plagiarism of programming exercise {} with title '{}' was successful in {}.", programmingExercise.get().getId(), programmingExercise.get().getTitle(),
-                formatDurationFrom(start));
+        plagiarismService.savePlagiarismResult(result);
+
+        optionalPreviousResult.ifPresent(plagiarismService::deletePlagiarismResult);
```
this logic of saving the new result and deleting the old could be reused by moving it into the plagiarismService.

philippbauch replied:
good point

derLalla commented on src/main/webapp/app/exercises/shared/plagiarism/plagiarism-header/plagiarism-header.component.ts:
```diff
@@ -18,18 +20,40 @@ export class PlagiarismHeaderComponent {
+            .put<void>(
+                `${this.resourceUrl}/${this.comparison.id}/status`,
+                { status },
+                {
+                    observe: 'response',
+                },
```
```suggestion
                { observe: 'response' },
```

derLalla commented on src/main/webapp/app/exercises/shared/plagiarism/plagiarism-inspector/plagiarism-inspector.component.ts:
```diff
@@ -92,9 +92,48 @@ export class PlagiarismInspectorComponent implements OnInit {
+
+        switch (this.exercise.type) {
+            case ExerciseType.MODELING: {
+                console.log('modeling plagiarism result');
```
you might want to remove the logs here

derLalla commented on src/main/webapp/app/exercises/shared/plagiarism/plagiarism-sidebar/plagiarism-sidebar.component.html:
```diff
@@ -15,8 +15,8 @@
                     [ngClass]="{
-                        confirmed: comparison.status === 'confirmed',
-                        denied: comparison.status === 'denied'
+                        confirmed: comparison.status === 'CONFIRMED',
+                        denied: comparison.status === 'DENIED'
```
you should use the defined enum which is defined in PlagiarismStatus.ts

philippbauch replied:
I tried to do that, but apparently, I can't use the TypeScript enum in an Angular template. Do you know a way around that?

krusche replied:
you have to make it available in the Typescript component, using something like
`readonly CONFIRMED = CONFIRMED`
(it's really weird, but otherwise the template does not know that this exists), there are some other places with enums which use this mechanism as well

derLalla reviewed (changes requested):
I have left some comments in the code

FuchsDominik reviewed (changes requested):
Looks good except for the error we found while inspecting modeling exercising. Programming exercises and text exercises worked fine

FrankeLukas reviewed (changes requested):
The "Check Plagiarism" button should ideally only be visible to instructors (see programming exercises).

daniels98it reviewed (commented):
Tested in testing session. 
Seems working fine :rocket:, except for 'Modeling exercises' as already mentioned

nicolasruscher reviewed (changes requested):
worked fine except for the modeling plag. detection

philippbauch commented:
I fixed a small bug causing trouble with plagiarism detection for modeling exercises. I also changed the "Check Plagiarism" buttons to only be visible if the user is at least an instructor.

FrankeLukas reviewed (approved):
Confirmed that modelling exercises work now. Also tested with a tutor that the buttons are no longer displayed.

ivanchimeno reviewed (approved):
Works correctly in testing session!

coolchock reviewed (approved):
tested during testing session, works as expected

iozbeyli reviewed (approved):
Works as expected

derLalla reviewed (approved):
worked fine in the testing session on ts2

FrankeLukas reviewed (approved):
Seems like you dropped the `isAtLeastInstructor` checks from the buttons again? Imho that can be done in a follow-up, since it is not critical (a tutor cannot view results or trigger plagiarism detection).
